File: gateway/web/src/minigame.contract.test.ts
import { describe, expect, it } from "vitest";
import fixture from "../../../contracts/minigame-v1.json";
import type { MiniGameInput, MiniGameInstance } from "./types";

describe("portable minigame contract", () => {
  it("consumes the same player-safe v1 fixture as Go and Rust", () => {
    const instance = fixture.instance as MiniGameInstance;
    const input = fixture.input as MiniGameInput;
    expect(instance.protocol_version).toBe(1);
    expect(instance.runtime.phase).toBe("active");
    expect(input.action).toBe("submit");
    expect("answers" in instance.definition).toBe(false);
  });
});

File: internal/engine/minigame_contract_test.go
package engine

import (
	"encoding/json"
	"os"
	"testing"
)

func TestCanonicalMiniGameFixtureIsPortableAndPlayerSafe(t *testing.T) {
	raw, err := os.ReadFile("../../contracts/minigame-v1.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Instance MiniGameInstance `json:"instance"`
		Input    MiniGameInput    `json:"input"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatal(err)
	}
	if err := fixture.Instance.Validate(); err != nil {
		t.Fatal(err)
	}
	if fixture.Input.Action != "submit" || len(fixture.Instance.Definition.Answers) != 0 {
		t.Fatalf("fixture leaked answers or lost input: %+v", fixture)
	}
}
